crates/uniflight/src/lib.rs

```diff
@@ -0,0 +1,428 @@
+}
+
+impl<T> Default for Shared<T> {
+    fn default() -> Self {
+        Self {
+            slot: AsyncMutex::new(None),
+        }
+    }
+}
+
+/// `BroadcastOnce` consists of shared slot and notify.
+#[derive(Clone)]
+struct BroadcastOnce<T> {
+    shared: Weak<Shared<T>>,
+}
+
+impl<T> BroadcastOnce<T> {
+    fn new() -> (Self, Arc<Shared<T>>) {
+        let shared = Arc::new(Shared::default());
+        (
+            Self {
+                shared: Arc::downgrade(&shared),
+            },
+            shared,
+        )
+    }
+}
+
+// After calling BroadcastOnce::waiter we can get a waiter.
+// It's in WaitList.
+struct BroadcastOnceWaiter<K, T, F> {
+    func: F,
+    shared: Arc<Shared<T>>,
+
+    key: K,
+    mapping: SharedMapping<K, T>,
+}
+
+impl<T> std::fmt::Debug for BroadcastOnce<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "BroadcastOnce")
+    }
+}
+
+#[expect(
+    clippy::type_complexity,
+    reason = "The Result type is complex but intentionally groups related items for the retry pattern"
+)]
+impl<T> BroadcastOnce<T> {
+    fn try_waiter<K, F>(
+        &self,
+        func: F,
+        key: K,
+        mapping: SharedMapping<K, T>,
+    ) -> Result<BroadcastOnceWaiter<K, T, F>, (F, K, SharedMapping<K, T>)> {
+        let Some(upgraded) = self.shared.upgrade() else {
+            return Err((func, key, mapping));
+        };
+        Ok(BroadcastOnceWaiter {
+            func,
+            shared: upgraded,
+            key,
+            mapping,
+        })
+    }
+
+    #[inline]
+    const fn waiter<K, F>(shared: Arc<Shared<T>>, func: F, key: K, mapping: SharedMapping<K, T>) -> BroadcastOnceWaiter<K, T, F> {
+        BroadcastOnceWaiter {
+            func,
+            shared,
+            key,
+            mapping,
+        }
+    }
+}
+
+// We already in WaitList, so wait will be fine, we won't miss
+// anything after Waiter generated.
+impl<K, T, F, Fut> BroadcastOnceWaiter<K, T, F>
+where
+    K: Hash + Eq,
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = T>,
+    T: Clone,
+{
+    async fn wait(self) -> T {
+        let mut slot = self.shared.slot.lock().await;
+        if let Some(value) = (*slot).as_ref() {
+            return value.clone();
+        }
+
+        let value = (self.func)().await;
+        *slot = Some(value.clone());
+
+        self.mapping.lock().remove(&self.key);
+
+        value
+    }
+}
+
+impl<K, T> UniFlight<K, T>
+where
+    K: Hash + Eq + Clone,
+{
+    /// Creates a new `UniFlight` instance.
+    #[inline]
+    #[must_use]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Execute and return the value for a given function, making sure that only one
+    /// operation is in-flight at a given moment. If a duplicate call comes in, that caller will
+    /// wait until the original call completes and return the same value.
+    pub fn work<F, Fut>(&self, key: K, func: F) -> impl Future<Output = T>
```

Comment:
I think the key should be &K here, you don't want the key to be consumed by this API.